File: frontend/src/features/battle-plan/taskTypeFilter.ts
import type { TaskType } from '../../lib/api'

/**
 * Battle Plan Task Type filter rules, shared with Android.
 *
 * The saved filter holds the Task Type ids the user chose directly, plus `unset`.
 * A chosen type covers its whole branch: `coding` also matches `coding/ai/evals`.
 */

export function isUnderTaskType(type: TaskType, ancestor: TaskType): boolean {
  return type.name.startsWith(`${ancestor.name}/`)
}

/** Every filter key that matches: the chosen keys plus the ids of their descendants. */
export function taskTypeFilterCoverage(filter: string[], taskTypes: TaskType[]): Set<string> {
  const chosen = taskTypes.filter((type) => filter.includes(String(type.id)))
  const covered = new Set(filter)
  for (const type of taskTypes) {
    if (chosen.some((ancestor) => isUnderTaskType(type, ancestor))) covered.add(String(type.id))
  }
  return covered
}

/** The shallowest chosen ancestor that covers `type`, or null when `type` is not covered that way. */
export function coveringTaskType(type: TaskType, filter: string[], taskTypes: TaskType[]): TaskType | null {
  return taskTypes
    .filter((ancestor) => filter.includes(String(ancestor.id)) && isUnderTaskType(type, ancestor))
    .sort((a, b) => a.name.length - b.name.length)[0] ?? null
}

/**
 * Chooses or clears `key`. Choosing a parent absorbs any descendants chosen individually,
 * since the parent already covers them.
 */
export function toggleTaskTypeFilter(filter: string[], key: string, taskTypes: TaskType[]): string[] {
  if (filter.includes(key)) return filter.filter((value) => value !== key)
  const type = taskTypes.find((row) => String(row.id) === key)
  if (!type) return [...filter, key]
  const absorbed = new Set(taskTypes.filter((row) => isUnderTaskType(row, type)).map((row) => String(row.id)))
  return [...filter.filter((value) => !absorbed.has(value)), key]
}

File: android/app/src/main/java/com/timebox/android/ui/battleplan/TaskTypeFilter.kt
package com.timebox.android.ui.battleplan

import com.timebox.android.data.TaskType

/**
 * Battle Plan Task Type filter rules.
 *
 * The filter holds the Task Type ids the user chose directly. A chosen type covers its
 * whole branch, so choosing `coding` also shows tasks typed `coding/ai` or `coding/ai/evals`.
 */

internal fun TaskType.isUnder(ancestor: TaskType): Boolean = name.startsWith(ancestor.name + "/")

/** Every Task Type id the filter matches: the chosen ids plus their descendants. */
internal fun taskTypeFilterCoverage(filter: Set<String>, taskTypes: List<TaskType>): Set<String> {
    val chosen = taskTypes.filter { it.id.toString() in filter }
    return filter + taskTypes.filter { type -> chosen.any { type.isUnder(it) } }.map { it.id.toString() }
}

/** The shallowest chosen ancestor that covers [type], or null when [type] is not covered that way. */
internal fun coveringTaskType(type: TaskType, filter: Set<String>, taskTypes: List<TaskType>): TaskType? =
    taskTypes.filter { it.id.toString() in filter && type.isUnder(it) }.minByOrNull { it.name.length }

/**
 * Chooses or clears [id]. Choosing a parent absorbs any descendants chosen individually,
 * since the parent already covers them.
 */
internal fun toggleTaskTypeFilter(filter: Set<String>, id: String, taskTypes: List<TaskType>): Set<String> {
    if (id in filter) return filter - id
    val type = taskTypes.find { it.id.toString() == id } ?: return filter + id
    val absorbed = taskTypes.filter { it.isUnder(type) }.map { it.id.toString() }.toSet()
    return filter - absorbed + id
}
